DOCS Review extending docs for truthfulness (#273)
silverstripe/developer-docs: en/02_Developer_Guides/05_Extending/05_Injector.md

```php
use App\MyClient;
use SilverStripe\Core\Injector\Injector;

$object = Injector::inst()->create(MyClient::class);
```

Repeated calls to `create()` create a new object each time.

```php
use App\MyClient;
use SilverStripe\Core\Injector\Injector;

$object = Injector::inst()->create(MyClient::class);
$object2 = Injector::inst()->create(MyClient::class);

// resolves to false
$object === $object2;
```

Arguments can be passed to the constructor of the object being instantiated by passing them in to `create()`. The method takes a variable-length argument list so you can pass in as many arguments as you need to the constructor.

```php
use App\MyClient;
use SilverStripe\Core\Injector\Injector;

$object = Injector::inst()->create(MyClient::class, $arg1, $arg2);
```

### Singleton Pattern

The `Injector` API can be used for the [singleton pattern](https://en.wikipedia.org/wiki/Singleton_pattern) through `get()`. Unlike `create()` subsequent calls to `get` return the same object instance as the first call.

```php
use App\MyClient;
use SilverStripe\Core\Injector\Injector;

// Fetches MyClient as a singleton
$object = Injector::inst()->get(MyClient::class);

> $object2 = Injector::inst()->get(MyClient::class);

Reviewer: Might also be worth mentioning that you can call nicer API `MyClient::singleton()` if MyClient has the Injectable trait on it

Reviewer 2: That's mentioned in the section below which is explicitly about that trait. I'll add a reference to it here.

Reviewer 2: Done

[info]
The arguments passed in for the singleton's constructor will only take effect the first time the singleton is instantiated - after that, because it is a singleton and has therefore already been instantiated, the constructor arguments will be ignored.
[/info]

```php
use App\MyClient;
use SilverStripe\Core\Injector\Injector;

// sets up MyClient as a singleton
$object = Injector::inst()->get(MyClient::class, constructorArgs: [$arg1, $arg2]);
```

#### Prototype services are never singletons

It is possible to tell `Injector` to always instantiate a new object for a given service even if it's requested it as a singleton. This is done by setting the 'type' for a given service definition to "prototype" in yaml configuration like so:

```yml
SilverStripe\Core\Injector\Injector:
  App\MyClient:
    type: 'prototype'
```

```php
use App\MyClient;
use SilverStripe\Core\Injector\Injector;

// Instantiates new MyClient objects each time, even though this would normally fetch a singleton
$object = Injector::inst()->get(MyClient::class);
$object2 = Injector::inst()->get(MyClient::class);

// resolves to false
$object === $object2;
```

> This is particularly useful when a service is intended to be [declared as a dependency](#dependencies-and-properties) for some other class, and you don't want that dependency to be a singleton.

Reviewer: Move this bit to the top of the section which explains the "why" of why you'd use this. I think this context is important because my first reaction to this new section was "why would I use this over regular `::create()`"?

Reviewer 2: Done

use App\MyController;
use SilverStripe\Core\Injector\Injector;

$controller = Injector::inst()->get(MyController::class);
```

Would perform the following steps:

* Fetches a singleton of the `App\MyController` service

> * Assuming this is the first time a `App\MyController` singleton is being instantiated:

Reviewer: ```suggestion
* If there no existing `App\MyController` instance for the singleton:
```

Reviewer 2: I'll change this a bit to 
> If there no existing instance for the `App\MyController` singleton

This is because the singleton doesn't have to be an instance of `App\MyController`

Reviewer 2: done

>   * Instantiates the service as an object of the `App\MyController` class since no other was defined

Reviewer: ```suggestion
  * Instantiates the service as an instance of the `App\MyController` class
```

Reviewer: Also make these sorts of changes for the other 'assuming' bullet points below

Reviewer 2: done

  * Look through the `dependencies` for the `App\MyController` service and fetch a singleton of the `App\PermissionService` service
  * Assuming this is the first time a `App\PermissionService` singleton is being instantiated:
    * Instantiates the service as an object of the `App\RestrictivePermissionService` class since that class was defined
    * Look at the properties to be injected for the `App\PermissionService` service fetch a singleton of the `App\MySQLDatabase` service
    * Assuming this is the first time a `App\MySQLDatabase` singleton is being instantiated:
      * Instantiates the service as an object of the `App\MySQLDatabase` class since no other was defined

>         * Evaluates and passes in the values of the constants or environment variables `dbusername` and `dbpassword` as arguments to the constructor

Reviewer: Indentation

Reviewer 2: What do you want to be changed about the indentation?
Note that I have intentionally indented these bullet points - e.g. "assuming this is the first time it's instantiated" applies to the indented items below that. Once the indent is finished, we no longer require that assumption for the rest of the steps.

Reviewer: The `Evaluates and passes in the values ` line is what I was referring to, seems like it shouldn't indented

Reviewer 2: done
